text/0000-array-private-classes

```diff
@@ -0,0 +1,198 @@
+data structures must not be made visible through iterator functions like `ArrayKeys`,
+`ArrayValues` and `ArrayPairs`. These classes can be made private as they are
+only used as return types for `Array` functions `keys`, `values` and `pairs`.
+The return values for these functions are changed to the more general interface
+`Iterator`.
+
+A new interface `RewindableIterator` is defined to allow for rewindable iterators,
+like it is the case for `Array` `values`.
+
+This design principle is applied to the other collection classes that expose
+internals too like:
+
+* `List`
+* `Map`
+* persistent `Map`
+* `Vec`
+* persistent `Vec`
+* `Set`
+* `Itertools`
+
+# Motivation
+
+This change brings:
+
+- Applying the design principle of
+[hiding implementation details](https://en.wikipedia.org/wiki/Information_hiding)
+but offer a general and stable interface. Returning interfaces instead of concrete
+classes allows changing the implementation. Usually, one must return the most
+general type that fullfils the contract of the function (in the case of the
+functions discussed in this RFC, iteration).
+- Collections' functions `keys`, `values` and `pairs` definitions are made more
+general. Iterators implementation details are not public. Internal classes used
+by implementation like `*Keys`, `*Values` and `*Pairs` are now
+[opaque data types](https://en.wikipedia.org/wiki/Opaque_data_type). Generally,
+when using these collection classes, clients are not interested by the iterators
+implementation, but by the types these iterators return and that is provided by
+the generic parameters.
+- The generic return signature of these 3 iterating functions is simpler to
+understand for clients of collection classes.
+- Makes the standard library more simple by hiding 18 specialised classes from
+stdlib of which 3 are from `builtin`.
+- The interface `RewindableIterator` is added to create rewindable iterators
+(can be re-start from first value).
+
+This change remains compatible with the existing code base but for client code
+that is directly using the classes `*Keys`, `*Values` and `*Pairs`. A search on
+Github shows that the impact is very limited.
+
+# Detailed design
+
+Iterating functions in collections `keys`, `values` and `pairs` are changed to
+return `Iterator` and the classes that implement these iterators are made private.
+Here are the full implementation of these functions for the `Array` class (changes
+in other collection classes are identical).
+
+```pony
+  fun keys(): Iterator[USize]^ =>
+    """
+    Return an iterator over the indices in the array.
+    """
+    _ArrayKeys[A, this->Array[A]](this)
+
+  fun values(): RewindableIterator[this->A]^ =>
+    """
+    Return an iterator over the values in the array.
+    """
+    _ArrayValues[A, this->Array[A]](this)
+
+  fun pairs(): Iterator[(USize, this->A)]^ =>
+    """
+    Return an iterator over the (index, value) pairs in the array.
+    """
+    _ArrayPairs[A, this->Array[A]](this)
```

Comment:
It strikes me as very weird that keys and pairs aren't currently rewindable. It seems like they should be rewindable. I think that rather than making this RFC a straight translation of what currently exists, that detailing changed like "making ArrayKeys and ArrayPairs" rewindable would be a good addition. 

The lack of rewindability of the existing keys and pairs (that probably happens elsewhere as well) seems more like an accident of history than a design choice.

Looking for oddities like this is why I'd like an enumeration of all the changes.

Reply:
I agree that those standard library iterators that *can* be made efficiently rewindable should all be made such.

Reply:
And I think the fact that you were able to notice that so quickly in the code actually speaks to one of the strengths of this approach - when I see the type name `ArrayValues` or `ArrayKeys` in the return signature I don't immediately know what features it exposes (I wouldn't even necessarily know it was an iterator at all just by the name). Whereas with the approach advocated by this RFC, as long as I know what `Iterator` and `RewindableIterator` are, I can immediately understand what it is and what features it exposes (as long as I already know about `Iterator` and/or `RewindableIterator`)

Reply:
DONE